### ISSUE TITLE
Problem with parallel requests

### ISSUE THREAD

**Issue by tompson** [p2/OAuth2#79]
When the access token expires and multiple requests happen, that all request a new access token via the refresh token, at the same time, some of them fail (because the access token from the first refresh token request does not work after the second refresh token request).

Parallel requests can be triggered from different parts of the app. For example when the app is launched the current location is sent to the server and a list of events is downloaded.

What would be the best way to make sure that no second refresh token request is made while the first is still running?

**Comment by p2** [comment]
The simplest way would be to create a "loader" class as a singleton (or otherwise reusing the same instance) that performs all your loading. This way you can implement an `isAuthorizing` flag and not perform another authorization if one is running.

On the develop branch I have actually added a private `isAuthorizing` Bool to the base class. It's currently only used in case authorization is aborted. It could easily be used to cancel an `authorize()` call in case one is already ongoing. This should make sense, what do you think? You could try to check for `isAuthorize` here: https://github.com/p2/OAuth2/blob/develop/Sources/Base/OAuth2.swift#L207

**Comment by tompson** [comment]
I guess the flag alone is not enough, I am currently trying to use `NSOperationQueue` like this

    lazy var authQueue: NSOperationQueue = {
        let queue = NSOperationQueue()
        queue.maxConcurrentOperationCount = 1
        return queue
    }()

    func authorize(callback: (wasFailure: Bool, error: ErrorType?) -> Void) {
        if oauth2.hasUnexpiredAccessToken() {
            callback(wasFailure: false, error: nil)
        } else {
            authQueue.addOperationWithBlock { () -> Void in
                let authsema : dispatch_semaphore_t = dispatch_semaphore_create(0)
                self.oauth2.afterAuthorizeOrFailure = { (wasFailure: Bool, error: ErrorType?) -> Void in
                    callback(wasFailure: wasFailure, error: error)
                    dispatch_semaphore_signal(authsema)
                }
                self.oauth2.authorize(params: nil, autoDismiss: true)
                dispatch_semaphore_wait(authsema, DISPATCH_TIME_FOREVER)
            }
        }
    }

the request I want to do in the first place is made in the callback closure.

**Comment by p2** [comment]
Yes, depends on if you want to immediately call the callback, with an error, when `authorize()` is already running, or if you want to run them sequentially, meaning the second and later calls will just use the token that was just obtained. Your solution should work that way, let me know how that works for you!

**Comment by tompson** [comment]
There seems to be still some kind of problem with my approach, needs some more testing and debugging today. 

For my case it is important that all the triggered requests are performed and no request is canceled because of a missing token.

**Comment by p2** [comment]
The authorize call does not take callbacks and always calls the same callback block, defined per OAuth2 instance. Hence I'm not convinced your proposed approach makes sense, it's just calling the same block multiple times.

I think a better approach would be to call your own `ready()` method from the `authorize()` callback, with that ready-method signaling that it's now okay to make requests.

**Comment by tompson** [comment]
hm, so setting `self.oauth2.afterAuthorizeOrFailure` to a new callback before calling `authorize()` does not work?

**Comment by p2** [comment]
That does work, but if an `authorize()` callback is already running and you're changing the callback, the original callback will not be called.

This is intentional. You will want one point in the app that retrieves the access token, because most likely you'll have user-interaction. That part should be independent from the requests you make.

**Comment by tompson** [comment]
The initial getting of the access token works fine for me, I have a separate view and I am explicitly calling `authorize()`

Where I have problems is when the access token expires and the user opens the app again. Then multiple request are sent in parallel to the server and the all need to get a new access token using the refresh token, no user interaction needed.

**Comment by p2** [comment]
You can call `authorize()` from the `applicationDidBecomeActive:` method, setting a flag or otherwise notifying your requests-sending component.

**Comment by tompson** [comment]
you mean using the flag to make sure the request are queued and sent after the `authorize()` call was finished?

**Comment by p2** [comment]
Yes, something along that path.

I'm making this up on the spot, but for example you create a `Loader` class; could be a singleton or AppDelegate holds on to it. This Loader class holds on to an `OAuth2XY` instance and could have a method `performRequest(request: NSURLRequest, ...)`. Now, whenever you call this method with a request, it checks `oauth2.hasUnexpiredAccessToken()` (or just executes the request and looks for a 401), and if it's `false`, remembers the request+callback, then calls `authorize()`. In the authorize callback, you tell the Loader that it's done and now the loader can work through all the requests it has enqueued.

Something along those lines I think could work. Maybe that's something that should be added to the framework.

**Comment by tompson** [comment]
ok, I made quite some progress here. now I am experiencing a strange issue: sometimes when the app is launched, it looks like the tokens are not read from the keychain:

    2016/01/19 01:40:24:066  === APPLICATION STARTED ===
    2016/01/19 01:40:24:077  OAuth2: Looking for items in keychain
    2016/01/19 01:40:24:079  OAuth2: Initialization finished
    2016/01/19 01:40:24:079  has no refresh token
    2016/01/19 07:56:46:116  Start synchronising termine
    2016/01/19 07:56:46:153  OAuth2: No access token, maybe I can refresh
    2016/01/19 07:56:46:153  OAuth2: I don't have a refresh token, not trying to refresh
    2016/01/19 07:56:46:153  OAuth2: No access token, requesting a new one
    2016/01/19 07:56:46:153  OAuth2: No username
    2016/01/19 07:56:46:154  authorization wasFailure Optional(No username)
    2016/01/19 07:56:46:158  OAuth2: No access token, maybe I can refresh
    2016/01/19 07:56:46:159  OAuth2: I don't have a refresh token, not trying to refresh
    2016/01/19 07:56:46:159  OAuth2: No access token, requesting a new one
    2016/01/19 07:56:46:159  OAuth2: No username
    2016/01/19 07:56:46:159  authorization wasFailure Optional(No username)

after killing the app and starting it again it works:

    2016/01/19 07:57:03:082  === APPLICATION STARTED ===
    2016/01/19 07:57:03:101  OAuth2: Looking for items in keychain
    2016/01/19 07:57:03:104  OAuth2: Found access token but it seems to have expired
    2016/01/19 07:57:03:104  OAuth2: Found refresh token
    2016/01/19 07:57:03:105  OAuth2: Initialization finished

**Comment by p2** [comment]
That's odd, especially since the log starts with "Looking for items in keychain".

**Comment by tompson** [comment]
I found the reason for this: the default mode in wich `SwiftKeychain` saves stuff to the Keychain is `kSecAttrAccessibleWhenUnlocked` which means they can not be read when the device is locked. Since my app needs to access them when doing a refresh in the background I need to change this somehow.

The constructor of the `SwiftKeychain` class `Keychain' would allow that:

    public init(serviceName name: String, accessMode: NSString = kSecAttrAccessibleWhenUnlocked, group: String? = nil) {
       
       self.accessMode = accessMode
       
       serviceName = name
       accessGroup = group
    }

@p2 should we change the default mode or make it configurable in p2.OAuth2?

**Comment by tompson** [comment]
I raised a separate issue #92

**Comment by tompson** [comment]
fixed with pull request #95

**Comment by carmof** [comment]
I might be late at this issue but I'd like to share my way of fixing this problem.

First I have all valid Tokens stored at the back-end, that's important. Each token is stored by the user's ID.
Ex: userID -> 'token', 'refresh token'

Whenever the "/refresh" endpoint receives a request it retrieves the tokens stored and checks the main token validity, if it's still valid the endpoint will return the stored tokens, otherwise it will create new ones, store them, and then return them as a response.

This way if there are two parallel requests to refresh the token the first to arrive will refresh it (create new ones), because the token is not valid, and return them as a response. And the second request will check the validity of the tokens that were just stored by the first request, it will see that they are still valid (because they were just created), and then it will return them instead of creating new ones.

Hope this helps.
